### DISCUSSION TITLE
Can wasmer be run inside another wasm VM?

### DISCUSSION THREAD

**Discussion by Zizico2** [wasmerio/wasmer#3174] [General]
### Summary

Can Wasmer run in Wasm? I'm developing a server app which would use webassembly for plugins, but the app itself could, eventually, be ran on wasm itself. Is this possible?

**Comment by jcaesar** [comment]
This is not possible with a normal wasmer binary. One reason is that wasmer comiles WASM binaries to native host CPU instructions. WASM offers no way of executing those compiled instructions in an instance's memory for obvious security reasons.

Possible alternatives: 
 * Use a self-hosting wasm engine, like wasm3.
 * Make a small wrapper binary that allows precisely what you need: wasm modules loading other wasm modules via wasmer. Your server app could then be a wasm module running in this wrapper binary.

**Comment by syrusakbary** [comment]
One small nit: Wasmer (Rust API) currently can run in other VM: the Javascript VM since Wasmer is fully compilable to JS via wasm-bindgen. But this is a trick of API translation rather than pushing into memory the compiled code.

As @jcaesar commented are two ways that Wasmer can run in other VMs:
* Implement a Wasm interpreter, so Wasmer can be self-hosted (like wasm3)
* Implement an ABI so Wasm modules can start other wasm modules, but for that other runtimes would also need to support this ABI (wasm-jit ABI?).
